rfcs/0070-introducing-decl-buffer.md

```diff
@@ -0,0 +1,210 @@
+physical memory. This makes analysis in TIR difficult and error-prone as one should always check
+whether a buffer in TIR is an implicitly-created one. 
+
+By introducing explicit `DeclBuffer` statement, we can require that a buffer must always be declared
+before any usages. This makes the creation and the usage of buffer better-managed within TIR.
+Developers (e.g pass writers) can collect buffer information such as allocation, aliasing by
+visiting `DeclBuffer` nodes.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`DeclBuffer` will be defined as 
+```
+class DeclBuffer : public Stmt {
+    Buffer buffer;  // the buffer declared
+    Stmt body;  // the scope of the buffer
+};
+```
+
+In TVM script, `T.buffer_decl` will be renamed to `T.decl_buffer` to make the name a verb phase that
+is consistent with the existing ones such as `T.alloc_buffer`, `T.match_buffer`. `T.decl_buffer`
+will be translated to a `DeclBuffer` object in TIR. This only changes the way parser handles
+`T.decl_buffer`, the user API of `T.decl_buffer` in TVM script will stay the same.
+
+In TIR, `DeclBuffer` will be handled in `StmtFunctor`. Visitors or mutators of `DeclBuffer` can be
+override to handle `DeclBuffer` in TIR passes.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Allocation of intermediate buffer
+The intermediate buffer inside `PrimFunc` can be declared and allocated in the following way:
+
+```
+Allocate {
+  data: A_data(Var(name=...))
+  extent: ...
+  body: DeclBuffer {
+    buffer: A(data=A_data, dtype=..., shape=...),
+    body: {
+      ...
+    }
+  }
+}
+```
+
+Alternatively, a buffer can be declared first and then allocated. In this case, any usage of the
+buffer before it is allocated is not allowed.
+
+## Declaration of buffer alias
+Buffer declared in `DeclBuffer` can reuse data variable from another buffer. This creates a buffer
+alias.
+
+```
+DeclBuffer {
+  buffer: A(data=Var(name=...), dtype=..., shape=...),
+  body: {
+    DeclBuffer {
+      buffer: A_alias(data=A.data, ...)
+      body: ...
+    }
+  }
+}
+```
+
+## Replace `preflattened_buffer_map` with buffer alias
+
+Currently, `PrimFunc` has two maps, `preflattened_buffer_map` and `buffer_map`, to specify the input
+buffer shapes. Before the flattening passes (`FlattenBuffer` and `StorageFlatten`),
+`preflattened_buffer_map` is empty and `buffer_map` contains the logical shapes of the buffers.
+After flattening, the logical shapes are moved to `preflattened_buffer_map`, and `buffer_map` will
+store the physical shapes of the buffers. The change of the information stored in `buffer_map` can
+be confusing. These two maps can be unified into a single `buffer_map` that defines the logical
+shapes of the input buffers. The buffer access in physical shape, which is an internal behavior of
+`PrimFunc` after flattening, can be achieved by using `DeclBuffer` to create buffer aliases in
```

Comment:
will there be an easy way to do this implemented as part of this RFC? @Lunderberg mentioned something about a utility function to at least compute the flattened buffer_map.

Reply:
@Lunderberg has a prototype for this https://github.com/apache/tvm/pull/10940, it is based on implicit buffer aliasing. This paragraph is suggesting moving to explicit buffer aliasing using `DeclBuffer`

Reply:
That's correct, the buffer flattening would be done through buffer aliasing.  After https://github.com/apache/tvm/pull/10940, the use of N-d/flattened buffers would be as below, depending on where you are in the lowering flow.

1. Before `BufferFlatten`/`FlattenStorage`: Buffers are declared in the `buffer_map`, and are not flattened.  Buffer access is done using N-d unflattened indices.
2. After `BufferFlatten`/`FlattenStorage`, but before `MakePackedAPI`: Buffers are declared in the `buffer_map`, and are not flattened.  Buffer access is done through a buffer alias, where the alias shares the same data pointer, but has a flattened shape and is accessed with flattened indices.
3. After `MakePackedAPI`: The `buffer_map` is empty.  Declarations of flattened buffers are done using the handles extracted using `tvm_struct_get`.  These flattened buffers are accessed with flattened indices.

As I understand it, this RFC would impact the TIR during parts 2 and 3.  Step 2 would have an explicit `DeclBuffer` to mark the aliasing, rather than an implicit aliasing based on re-used `Var Buffer::data`.  Step 3 would have an explicit `DeclBuffer` to mark the use of the `T.handle` in a buffer, rather than being implicit in the first use.

Reply:
thanks for elaborating that! could we add these to the RFC?